File: core/src/main/java/io/ably/lib/types/Channels.java
package io.ably.lib.types;


/**
 * A collection of Channels associated with an Ably instance.
 */
public interface Channels<Channel extends AblyChannel> extends ReadOnlyMap<String, Channel> {
    /**
     * Get the named channel; if it does not already exist,
     * create it with default options.
     *
     * @param channelName the name of the channel
     * @return the channel
     */
    Channel get(String channelName);

    /**
     * Get the named channel and set the given options, creating it
     * if it does not already exist.
     *
     * @param channelName    the name of the channel
     * @param channelOptions the options to set (null to clear options on an existing channel)
     * @return the channel
     * @throws AblyException
     */
    Channel get(String channelName, ChannelOptions channelOptions) throws AblyException;

    /**
     * Remove this channel from this Ably instance. This detaches from the channel
     * and releases all other resources associated with the channel in this client.
     * This silently does nothing if the channel does not already exist.
     *
     * @param channelName the name of the channel
     */
    void release(String channelName);

    int size();

    Iterable<Channel> values();
}

File: pubsub-adapter/src/main/kotlin/com/ably/pubsub/Channels.kt
package com.ably.pubsub

import io.ably.lib.realtime.ChannelState
import io.ably.lib.types.ChannelOptions

/**
 * Represents collection of managed Channel instances
 */
interface Channels<ChannelType> : Iterable<ChannelType> {

  /**
   * Checks if channel with specified name exists
   * <p>
   * Spec: RSN2, RTS2
   * @param name The channel name.
   * @return `true` if it contains the specified [name].
   */
  fun contains(name: String): Boolean

  /**
   * Creates a new [Channel] object, or returns the existing channel object.
   * <p>
   * Spec: RSN3a, RTS3a
   * @param name The channel name.
   * @return A [Channel] object.
   */
  fun get(name: String): ChannelType

  /**
   * Creates a new [Channel] object, with the specified [ChannelOptions], or returns the existing channel object.
   * <p>
   * Spec: RSN3c, RTS3c
   * @param name The channel name.
   * @param options A [ChannelOptions] object.
   * @return A [Channel] object.
   */
  fun get(name: String, options: ChannelOptions): ChannelType

  /**
   * Releases a [Channel] object, deleting it, and enabling it to be garbage collected.
   * It also removes any listeners associated with the channel.
   * To release a channel, the [ChannelState] must be `INITIALIZED`, `DETACHED`, or `FAILED`.
   * <p>
   * Spec: RSN4, RTS4
   * @param name The channel name.
   */
  fun release(name: String)
}
